### PR TITLE
[total_daily_energy] Replace loop() with timeout-based midnight reset

### PR DESCRIPTION
**Pull request by bdraco**
# What does this implement/fix?

The `total_daily_energy` component previously ran its `loop()` every main loop iteration (~130Hz) just to check if midnight had passed. This called `time_->now()` on every iteration to construct a full `ESPTime` struct, making it the most expensive component by call count (7804 calls/60s and 423.8ms total CPU time per minute).

This replaces the `loop()` with a `set_timeout` that schedules the midnight reset at the exact right time. The timeout is (re)scheduled:
- At `setup()` if time is already valid
- On every NTP sync callback (in case the clock jumped across midnight)
- After each midnight reset fires (self-rescheduling)

### Runtime stats (ESP8266)

**Before (with `loop()`)** — `total_daily_energy` was the #1 most expensive component:
```
total_daily_energy.sensor: count=7804, avg=0.054ms, max=0.24ms, total=423.8ms
api:                       count=7804, avg=0.007ms, max=1.55ms, total=51.1ms
mdns:                      count=1121, avg=0.037ms, max=0.09ms, total=41.5ms
```

**After (timeout-based)** — `total_daily_energy` no longer appears (zero loop calls):
```
mdns:                      count=1090, avg=0.060ms, max=2.50ms, total=65.5ms
api:                       count=7580, avg=0.008ms, max=2.75ms, total=60.0ms
wifi:                      count=7580, avg=0.007ms, max=4.66ms, total=51.6ms
```

~424ms of CPU time per minute eliminated (ESP8266 is single-core, so wall clock ≈ CPU time). For context, a bluetooth proxy actively processing BLE advertisements uses ~208ms/min — a no-op midnight check was **2x more expensive** than real BLE proxy work.

### DST handling

`set_timeout` counts real elapsed milliseconds (`millis()`), not wall clock time. DST transitions shift wall clock by 1 hour without changing `millis()`, which would cause a single timeout to fire up to 1 hour late (spring-forward) or early (fall-back).

To handle this, `schedule_midnight_reset_()` uses a two-stage approach:
- **>90 minutes from midnight:** schedules a pre-check at midnight minus 90 minutes
- **≤90 minutes from midnight:** schedules the actual reset with a +1s buffer

The pre-check recalculates with fresh wall clock time, so even after a 1-hour DST shift the final timeout is accurate. DST transitions typically occur at 2 AM, well outside the 90-minute window.

NTP sync also triggers rescheduling via `add_on_time_sync_callback`, but DST transitions do not fire this callback (DST changes local time interpretation, not the epoch).

A standalone DST simulation test (`dst_test.cpp`) is attached to this PR, verifying correctness for spring-forward, fall-back, edge cases (90-min boundary, exactly midnight, late night), and self-correcting behavior when DST causes early/late pre-check firing.

## Flash impact

+576 bytes on ESP8266. Breakdown:

- **One-time template costs (shared with other components):** ~234 bytes
  - `Callback<void()>` `_M_realloc_insert` (198 bytes) — first ESP8266 user of `add_on_time_sync_callback` that triggers this instantiation. Will be eliminated entirely by #15272.
  - `Scheduler::set_timeout(uint32_t)` (36 bytes) — first user of this overload.
- **Per-component logic:** ~342 bytes
  - `schedule_midnight_reset_()` (292 bytes), lambda wrappers (~50 bytes)
- **Removed:** `loop()` (160 bytes)

After #15272 merges, the net cost drops to ~378 bytes — a reasonable tradeoff for eliminating ~7800 unnecessary `time_->now()` calls per minute.

## Types of changes

- [ ] Bugfix (non-breaking change which fixes an issue)
- [ ] New feature (non-breaking change which adds functionality)
- [ ] Breaking change (fix or feature that would cause existing functionality to not work as expected) — [policy](https://developers.esphome.io/contributing/code/#what-constitutes-a-c-breaking-change)
- [ ] Developer breaking change (an API change that could break external components) — [policy](https://developers.esphome.io/contributing/code/#what-is-considered-public-c-api)
- [ ] Undocumented C++ API change (removal or change of undocumented public methods that lambda users may depend on) — [policy](https://developers.esphome.io/contributing/code/#c-user-expectations)
- [x] Code quality improvements to existing code or addition of tests
- [ ] Other

**Related issue or feature (if applicable):**

n/a

**Pull request in [esphome-docs](https://github.com/esphome/esphome-docs) with documentation (if applicable):**

n/a

## Test Environment

- [x] ESP32
- [x] ESP32 IDF
- [x] ESP8266
- [ ] RP2040/RP2350
- [ ] BK72xx
- [ ] RTL87xx
- [ ] LN882x
- [ ] nRF52840

## Example entry for `config.yaml`:

```yaml
# No config changes needed
```

## Checklist:
  - [x] The code change is tested and works locally.
  - [ ] Tests have been added to verify that the new code works (under `tests/` folder).

If user exposed functionality or configuration variables are added/changed:
  - [ ] Documentation added/updated in [esphome-docs](https://github.com/esphome/esphome-docs).

### CODE REVIEW

File: esphome/components/total_daily_energy/total_daily_energy.cpp [hunk 1]
```diff
@@ -15,28 +21,44 @@ void TotalDailyEnergy::setup() {
...
+
+  ESP_LOGD(TAG, "Scheduling midnight reset in %us", seconds_until_midnight);
+  this->set_timeout(MIDNIGHT_TIMEOUT, seconds_until_midnight * MS_PER_SECOND, [this]() {
+    this->reset_day_(0);
+    this->schedule_midnight_reset_();
+  });
+}
+
+void TotalDailyEnergy::reset_day_(uint16_t day_of_year) {
+  this->last_day_of_year_ = day_of_year;
```
[the remark this answers is not in the document]
Comment (bdraco) [reply esphome/components/total_daily_energy/total_daily_energy.cpp]: its already imprecise because callback time is not zero, time to read is not zero, overhead is not zero. We can't calculate that. Its the same issue as before and at most its one misattributed read

**Comment by bdraco** [comment]
[dst_test.zip](https://github.com/user-attachments/files/26477147/dst_test.zip)

**Review by clydebarrow — APPROVED** [review]

**Comment by bdraco** [comment]
thanks
